### PR TITLE
Chat hook

### PR DESCRIPTION
**Pull request by howawong**
1. Add Hook decoration
2. Call lambdas after operations
3. Query users in calling hooks

Connects #158

### CODE REVIEW

**Comment by howawong** [comment]
Sample Python Code
```python
def register_custom_hooks(settings):
    @after_conversation_created
    def after_conversation_created_hook(conversation, participants):
        print("diu")
        print(conversation)
        print(participants)

    @after_users_added_to_conversation
    def after_users_added_to_conversation_hook(conversation, participants, new_users):
        print("new users", new_users)

    @after_users_removed_from_conversation
    def after_users_removed_from_conversation_hook(conversation, participants, old_users):
        print("old users", old_users)

    @after_conversation_deleted
    def after_conversation_deleted_hook(conversation, participants):
        print("deleted", participants)

    @after_conversation_updated
    def after_conversation_updated_hook(conversation, participants):
        print("updated", participants)

    @after_message_deleted
    def after_message_deleted_hook(message, conversation, participants):
        print(message)
        print(conversation)
        print(participants)
        print("deleted")

    @after_message_updated
    def after_message_updated_hook(message, conversation, participants):
        print(message)
        print(conversation)
        print(participants)
        print("updated")

    @after_message_sent
    def after_message_sent_hook(message, conversation, participants):
        print(message)
        print(conversation)
        print(participants)
        print("sent")

    @typing_started
    def typing_started_hook(conversation, participants, events):
        print(participants)
        print(conversation)
        print(events)
        print("typing started")

```

File: chat/hooks.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+                         AFTER_USERS_REMOVED_FROM_CONVERSATION_HOOK,
+                         TYPING_STARTED_HOOK)
+from .predicate import Predicate
+from .query import Query
+from .utils import _get_container
+
+
+def __send_hook(name, params):
+    container = _get_container()
+    print(name)
```
Comment (rickmak) [inline chat/hooks.py]: use log.

**Comment by rickmak** [comment]
Seem we can support both with or without `()`. Let do the support later.

**Comment by howawong** [comment]
FYI, If we need to handle bracket case, then just follow https://github.com/pytest-dev/pytest/blob/2e04771893f17b571e6ab48f6884a6a70cde8fc8/_pytest/python.py#L132 and  https://stackoverflow.com/questions/35572663/using-python-decorator-with-or-without-parentheses, without asking end user to upgrade their codes.

**Pull request merged**